cmd/holo/internal/generators.go

```diff
@@ -0,0 +1,173 @@
+
+package impl
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+)
+
+// RunGenerators executes all generators in the generator directory
+// and changes the resource path of plugins for which files were
+// generated to.
+func RunGenerators(config *Configuration) error {
+	inputDir := getGenertorsDir()
+	if _, err := os.Stat(inputDir); err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
+	targetDir, err := getGeneratorCacheDir()
+	if err != nil {
+		return fmt.Errorf(
+			"couldn't access cache-dir ('%s') for generators: %s",
+			targetDir, err,
+		)
+	}
+	fmt.Fprintln(Stdout, "Processing generators...")
+	filepath.Walk(inputDir,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				Warnf(Stderr, "%s: %s", path, err.Error())
+				return nil
+			}
+			if isExecutableFile(info) {
+				fmt.Fprintln(Stdout, path)
+				if err = runGenerator(path, targetDir); err != nil {
+					Errorf(
+						Stderr,
+						"Running generator failed: %s", err.Error(),
+					)
+				}
+			}
+			return nil
+		})
+	for _, plugin := range config.Plugins {
+		if err := updatePluginPaths(plugin, targetDir); err != nil {
+			Errorf(Stderr,
+				"Failed to perpare generated dir for plugin '%s': %s",
+				plugin.id, err.Error(),
+			)
+		}
+	}
+	return nil
+}
+
+func updatePluginPaths(plugin *Plugin, dir string) error {
+	pluginDir := plugin.ResourceDirectory()
+	newPluginDir := filepath.Join(dir, plugin.id)
+	if info, err := os.Stat(newPluginDir); err == nil && info.IsDir() {
+		// Files were generated for this plugin.
+		// Fill the plugins directory with existsing static files.
+		if err := symlinkFiles(pluginDir, newPluginDir); err != nil {
+			return err
+		}
+		// Change the plugin resource dir to point to the generated dir.
+		resource, _ := filepath.Rel(RootDirectory(), dir)
+		plugin.SetResourceRoot(resource)
+	}
+	return nil
+}
+
+func symlinkFiles(oldDir string, newDir string) error {
+	return filepath.Walk(oldDir,
+		func(oldFile string, info os.FileInfo, err error) error {
+			if err != nil || oldFile == oldDir {
+				return err
+			}
+			relPath, _ := filepath.Rel(oldDir, oldFile)
+			newFile := filepath.Join(newDir, relPath)
+			err = os.Symlink(oldFile, newFile)
+			if os.IsExist(err) {
+				// newFile already exists. Examine it.
+				newFileInfo, err := os.Lstat(newFile)
+				if err == nil && info.IsDir() && !newFileInfo.IsDir() {
+					// newFile exists but is not a directory.
+					// If oldFile is a directory trying to symlink its contents
+					// will result in errors. Skip it.
+					return filepath.SkipDir
+				}
+				return nil
+			}
+			if err != nil {
+				return err
+			}
+			if info.IsDir() {
+				// Symlink to dir was created don't check its contents.
+				return filepath.SkipDir
+			}
+			return nil
+		})
+}
+
+func runGenerator(fileToRun string, targetDir string) error {
+	cmd := exec.Command(fileToRun)
+	env := os.Environ()
+	env = append(
+		env,
+		fmt.Sprintf("OUT=%s", targetDir),
```

Comment:
Generators should also get `HOLO_RESOURCE_DIR` and `HOLO_CACHE_DIR` to ensure that they don't hardcode `/usr/share/holo` or `/tmp`.

Reply:
`HOLO_RESOURCE_DIR` is usually adjusted to be plugin specific. What would be the content of it for generators? Correct me if I am wrong, but isn't `/use/share/holo` currently hard coded in plugin.go? In other words the plugins are flexible to where their resource files live, but holo isn't. We should potentially introduce a `HOLO_RESOURCE_ROOT`.